src/main/scala/chiseltest/internal/BackendInterface.scala

```diff
@@ -59,43 +42,80 @@ trait BackendInterface {
     stale:   Boolean
   ): Unit
 
+  // Optional API
+  /** Writes a value to a clock.
+    *
+    * This is not supported by most of cycle-based simulators.
+    */
+  def pokeClock(signal: Clock, value: Boolean): Unit = {
+    throw new PokeClockImplementedException(s"poke to clock not implemented in ${getClass.getName}")
+  }
+
+  /** Read the value of a clock.
+    *
+    * This is not supported by most of cycle-based simulators.
+    */
+  def peekClock(signal: Clock): Boolean = {
+    throw new PeekClockImplementedException(s"peek to clock not implemented in ${getClass.getName}")
+  }
+
   /** Sets the timeout of the clock: the number of cycles the clock can advance without
     * some non-nop poke operation.
     * Setting cycles=0 disables the timeout.
     * Setting cycles=1 means every cycle must have some non-nop poke operation.
     * Resets the idle counter associated with the specified clock.
+    *
+    * TODO:
+    *   consider deprecating This API.
+    *   This consumes too much overhead to maintain a timeout function,
+    *   a better solution is adding Annotation and FIRRTL pass to add the timeout circuit,
+    *   let simulator to handle timeout.
     */
-  def setTimeout(signal: Clock, cycles: Int): Unit
+  def setTimeout(signal: Clock, cycles: Int): Unit = {
+    throw new ClockTimeoutNotImplementedException(s"setTimeout not implemented in ${getClass.getName}")
+  }
 
-  /** Advances the target clock by one cycle.
-    */
-  def step(signal: Clock, cycles: Int): Unit
+  // Threading API
+  def doFork(runnable: () => Unit, name: Option[String], region: Option[Region]): AbstractTesterThread = {
+    throw new ThreadNotImplementedException(s"fork/join not implemented in ${getClass.getName}")
+  }
```

Comment:
By default `ThreadNotImplementedException` will be thrown if not using threading API, while user use `fork/join`